antipackage.py

```diff
@@ -1,130 +1,401 @@
 anytime they change on GitHub.
 """
 
 from __future__ import print_function
+
 import os
 import sys
-from urllib import urlretrieve
-import hashlib
+import shutil
+import json
+import zipfile
 import shutil
 
-class InstallError(Exception):
-    pass
+try:
+    # python 2 import
+    from urllib import urlretrieve
+except ImportError:
+    # python 3 import
+    from urllib.requests import urlretrieve
 
-class GitHubImporter(object):
-    
-    def __init__(self):
-        self.base_dir = os.path.expanduser('~/.antipackage')
-        if not os.path.exists(self.base_dir):
-            os.makedirs(self.base_dir)
-        sys.path.append(self.base_dir)
-
-    def _parse_fullname(self, fullname):
-        comps = fullname.split('.')
-        top, username, repo, modname = None, None, None, None
-        if len(comps)>=1:
-            top = 'github'
-        if len(comps)>=2:
-            username = comps[1]
-        if len(comps)>=3:
-            repo = comps[2]
-        if len(comps)>=4:
-            modname = comps[3]
-        return top, username, repo, modname
+from IPython.utils.tempdir import TemporaryDirectory
+
+class RetrieveError(Exception): pass
+class GitHubError(Exception): pass
+
+# - - - - - - - - -
+# Utility Functions
+# - - - - - - - - -
+
+_conv = {}
+
+def replace_char(key, value=None, remove=False):
+    """Replace `key` with `value` in all import calls
+
+    Notes
+    -----
+    setting `remove` to True deletes a previously
+    created replacement rule assigned to `key`"""
+    if not remove:
+        if key != '.':
+            _conv[key] = value
+        else:
+            raise ValueError("'.' is a reserved character")
+    else:
+        del _conv[key]
+
+def repr_conv(chars):
+    """Applies replacements specified by `replace_char`"""
+    for key in _conv:
+        chars = chars.replace(key, _conv[key])
+    return chars
+
+def display_download(count, block, size):
+    """Displays download progress for sizable files
+
+    Notes
+    -----
+    Supply as the 'reporthook' for urlretrieve"""
+    fraction = size/block
+    if fraction>10:
+        percent = int(count*100/fraction)
+        sys.stdout.write("\rDownload in progress... %d%%" % percent)
+        sys.stdout.flush()
+
+# - - - - - - - - - - - - - - - -
+# Setup Base Dir and Pinning File
+# - - - - - - - - - - - - - - - -
+
+def _setup(directory, fname):
+    base_path = os.path.expanduser('~/'+directory)
+    if not os.path.exists(base_path):
+        os.makedirs(base_path)
+    data_path = base_path+'/'+fname
+    if not os.path.isfile(data_path):
+        with open(data_path, 'w') as f:
+            json.dump({},f)
+    return base_path, data_path
+
+_BASE_DIR, _DATA_PATH = _setup('.antipackage','pinnings.json')
+
+# - - - - - - - - - - - - - - - - - - - - -
+# Main Data Management Object and Functions
+# - - - - - - - - - - - - - - - - - - - - -
+
+def pin(path, **kwargs):
```

Comment:
The signature of this function should be `pin(path, sha=None, branch=None, tag=None)`